### PR TITLE
Performance improvements + Java style improvements

### PR DESCRIPTION
**Pull request by dstepanov**
Some performance improvements:
- MethodLookup is now instantiating the services, it should be faster and avoid reflection caches
- Detected some typecast problems in a basic HelloWorld
- Removed reflection uses to find the repeatable annotation container
- `BeanDefinitionProducer` to cache some of the `enabled` checks and the definition instance
- Introduced `CollectionUtils.newSet/../Map` helper methods to create set/map collection with an expected element count
- Reduced some of the streams and lambdas spotted in the profiled (might be a bit of controversial)
- Supplier memorize methods rewritten to use double-check locking, inspired by Guava

Styling in some classes to use the variable defined in the `instanceof`, eliminated some of the embeddings by using fast return.

### CODE REVIEW

File: core/src/main/java/io/micronaut/core/annotation/AnnotationValue.java [hunk 1]
```diff
@@ -231,20 +230,21 @@ Map<String, String> getProperties(@NonNull String member) {
...
-            if (CollectionUtils.isNotEmpty(values)) {
-                Map<String, String> props = new LinkedHashMap<>(values.size());
-                for (AnnotationValue<Annotation> av : values) {
-                    String name = av.stringValue(keyMember).orElse(null);
-                    if (StringUtils.isNotEmpty(name)) {
-                        av.stringValue(AnnotationMetadata.VALUE_MEMBER, valueMapper).ifPresent(v -> props.put(name, v));
-                    }
-                }
-                return Collections.unmodifiableMap(props);
+        if (StringUtils.isEmpty(member)) {
```
Comment (dstepanov) [inline core/src/main/java/io/micronaut/core/annotation/AnnotationValue.java:233]: Just style changes, no logic

File: core/src/main/java/io/micronaut/core/type/RuntimeTypeInformation.java [hunk 2]
```diff
@@ -39,29 +40,23 @@
...
 
     /**
      * Returns whether the annotation metadata specifies the type as single.
+     * @param type The return type
      * @param annotationMetadata The annotation metadata provider
      * @return True if does
      */
-    static boolean isSpecifiedSingle(AnnotationMetadataProvider annotationMetadata) {
-        for (TypeInformationProvider provider : TYPE_INFORMATION_PROVIDERS) {
+    static boolean isSpecifiedSingle(Class<?> type, AnnotationMetadataProvider annotationMetadata) {
```
Comment (dstepanov) [inline core/src/main/java/io/micronaut/core/type/RuntimeTypeInformation.java:55]: Added an extra check to avoid reactive check when the type is a basic Java type.

File: core/src/main/java/io/micronaut/core/util/StreamUtils.java [hunk 3]
```diff
@@ -189,7 +191,10 @@ public static <T, A extends Collection<T>> Collector<T, A, Collection<T>> toImmu
     /**
      * @param <T> The type
      * @return An immutable collection
+     * @deprecated use Stream#toList
      */
+    @Deprecated(forRemoval = true)
+    @NextMajorVersion("Remove after Micronaut 4 milestone 1")
     public static <T> Collector<T, Collection<T>, Collection<T>> toImmutableCollection() {
```
Comment (dstepanov) [inline core/src/main/java/io/micronaut/core/util/StreamUtils.java:198]: We should use `toList(`

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -175,8 +175,8 @@ jcache = { module = "javax.cache:cache-api", version.ref = "jcache" }
 
 jetty-alpn-openjdk8-client = { module = "org.eclipse.jetty:jetty-alpn-openjdk8-client", version.ref = "jetty" }
 
-jmh = { module = "org.openjdk.jmh:jmh-core", version.ref = "jmh" }
-jmh-generator-annprocess = { module = "org.openjdk.jmh:jmh-core", version.ref = "jmh" }
+jmh-core = { module = "org.openjdk.jmh:jmh-core", version.ref = "jmh" }
+jmh-generator-annprocess = { module = "org.openjdk.jmh:jmh-generator-annprocess", version.ref = "jmh" }
```
Comment (dstepanov) [inline gradle/libs.versions.toml:179]: I think that's correct dependency now

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyBinderRegistrar.java [hunk 5]
```diff
@@ -37,7 +37,7 @@
  * @author graemerocher
  * @since 2.0.0
  */
-@Singleton
+@Prototype
 @Internal
 class NettyBinderRegistrar implements BeanCreatedEventListener<RequestBinderRegistry> {
```
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyBinderRegistrar.java:42]: No need to be singleton as it's only loaded once

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/converters/NettyConverters.java [hunk 6]
```diff
@@ -47,13 +47,13 @@
  * @author graemerocher
  * @since 1.0
  */
-@Singleton
+@Prototype
 @Internal
 public class NettyConverters implements TypeConverterRegistrar {
```
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/converters/NettyConverters.java:52]: All type converters should be prototypes as they are only needed at the startup

File: http/src/main/java/io/micronaut/http/converters/HttpConverterRegistrar.java [hunk 7]
```diff
@@ -32,17 +33,17 @@
...
-    private final ResourceResolver resourceResolver;
+    private final Provider<ResourceResolver> resourceResolver;
 
     /**
      * Default constructor.
      *
      * @param resourceResolver The resource resolver
      */
-    protected HttpConverterRegistrar(ResourceResolver resourceResolver) {
+    protected HttpConverterRegistrar(Provider<ResourceResolver> resourceResolver) {
```
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/converters/HttpConverterRegistrar.java:46]: Avoid fetching it at the startup

File: inject-java/src/test/groovy/io/micronaut/inject/factory/nullreturn/NullReturnFactorySpec.groovy [hunk 8]
```diff
@@ -83,22 +80,20 @@ class Test2 {}
...
         expect: "1 is created for D"
         beanContext.getBeansOfType(D).size() == 1
         beanContext.getBean(D, Qualifiers.byName("one"))
-        factory.bCalls == 6
-        factory.cCalls == 4
         factory.dCalls == 2 //2 C beans
 
         and: "1 is created for D2"
-        beanContext.getBeansOfType(D2).size() == 1
+        beanContext.getBeansOfType(D2).size() == 1 // D two is disabled
```
Comment (dstepanov) [inline inject-java/src/test/groovy/io/micronaut/inject/factory/nullreturn/NullReturnFactorySpec.groovy:90]: Changes in `AbstractInitializableBeanDefinition` where `argument.isDeclaredNullabl` is added brought those changes, but I think it's the correct behaviour

File: inject/src/main/java/io/micronaut/context/AbstractInitializableBeanDefinition.java [hunk 9]
```diff
@@ -2130,6 +2141,9 @@ private <K> K resolveBean(
             boolean isNotInnerConfiguration = !precalculatedInfo.isConfigurationProperties || !isInnerConfiguration(argument);
             ConfigurationPath previousPath = isNotInnerConfiguration ? resolutionContext.setConfigurationPath(null) : null;
             try {
+                if (argument.isDeclaredNullable()) {
```
Comment (dstepanov) [inline inject/src/main/java/io/micronaut/context/AbstractInitializableBeanDefinition.java:2145]: This change requires the nullable factory test change

File: inject/src/main/java/io/micronaut/inject/BeanDefinition.java [hunk 10]
```diff
@@ -267,7 +267,7 @@ default <R> Stream<ExecutableMethod<T, R>> findPossibleMethods(String name) {
     /**
      * @return The {@link ExecutableMethod} instances for this definition
      */
-    default Collection<ExecutableMethod<T, ?>> getExecutableMethods() {
+    default Collection<ExecutableMethod<T, Object>> getExecutableMethods() {
```
Comment (dstepanov) [inline inject/src/main/java/io/micronaut/inject/BeanDefinition.java]: This signature makes more sense and allows us to aggregate methods without a compiler error

**Comment by graemerocher** [comment]
One failure

**Comment by graemerocher** [comment]
if you could address the sonar bugs as well https://sonarcloud.io/project/issues?id=micronaut-projects_micronaut-core&pullRequest=8586&resolved=false&types=BUG

**Comment by dstepanov** [comment]
@graemerocher PR adds more forgotten caches cleanup on the bean context stop. It detected a use case where Netty embedded server test would stop and start the context and because the environment is also reset, in this case, some of the beans didn't load because of the missing properties provided in the test. The solution is to copy the env during stop-start of the embedded context if it also stops, start the bean context.

**Review by graemerocher — APPROVED** [review]

File: benchmarks/build.gradle [hunk 11]
```diff
@@ -17,15 +18,14 @@ dependencies {
...
-    jmh libs.jmh.generator.annprocess
+    jmh libs.jmh.core
 }
 jmh {
     includes = ['io.micronaut.http.server.StartupBenchmark']
     duplicateClassesStrategy = DuplicatesStrategy.WARN
-    warmupIterations = 2
-    iterations = 3
-    fork = 1
+//    warmupIterations = 2
```
Comment (sdelamo) [inline benchmarks/build.gradle]: should we just remove it?

File: context/src/main/java/io/micronaut/runtime/context/env/ConfigurationIntroductionAdvice.java [hunk 12]
```diff
@@ -106,10 +106,11 @@ private Object resolveProperty(MethodInvocationContext<Object, Object> context,
         );
 
         if (defaultValue != null) {
-            return value.orElseGet(() -> environment.convertRequired(
```
Comment (sdelamo) [inline context/src/main/java/io/micronaut/runtime/context/env/ConfigurationIntroductionAdvice.java:109]: is there are reason why we prefer the `orElse` null and the if statement vs the `orElseGet`?

File: core/src/main/java/io/micronaut/core/io/file/DefaultFileSystemResourceLoader.java [hunk 13]
```diff
@@ -66,7 +66,7 @@ public DefaultFileSystemResourceLoader(String path) {
      * @param path The path
      */
     public DefaultFileSystemResourceLoader(Path path) {
-        this.baseDir = SupplierUtil.memoizedNonEmpty(() -> {
+        this.baseDir = SupplierUtil.memoized(() -> {
```
Comment (sdelamo) [inline core/src/main/java/io/micronaut/core/io/file/DefaultFileSystemResourceLoader.java:69]: is this change intentional ?

File: core/src/main/java/io/micronaut/core/type/Argument.java [hunk 14]
```diff
@@ -432,7 +436,7 @@ static <T> Argument<T> of(
         if (ArrayUtils.isEmpty(typeParameters)) {
             return of(type);
         }
-        return new DefaultArgument<>(type, NameUtils.decapitalize(type.getSimpleName()), AnnotationMetadata.EMPTY_METADATA, typeParameters);
+        return new DefaultArgument<>(type, null, AnnotationMetadata.EMPTY_METADATA, typeParameters);
```
Comment (sdelamo) [inline core/src/main/java/io/micronaut/core/type/Argument.java:439]: is this intentional ?

File: core/src/main/java/io/micronaut/core/type/DefaultArgument.java [hunk 15]
```diff
@@ -214,10 +228,13 @@ public Class<T> getType() {
     @Override
     @NonNull
     public String getName() {
-        if (name == null) {
-            return getType().getSimpleName();
+        if (name != null) {
+            return name;
+        }
+        if (namePrecalculated == null) {
+            namePrecalculated = NameUtils.decapitalize(type.getSimpleName());
```
Comment (sdelamo) [inline core/src/main/java/io/micronaut/core/type/DefaultArgument.java:235]: Before if name was null we were returning : 


```java
return getType().getSimpleName();
```

now we return 

```java
NameUtils.decapitalize(type.getSimpleName());
```

is this intentional?

File: core/src/main/java/io/micronaut/core/util/StreamUtils.java [hunk 16]
```diff
@@ -189,7 +191,10 @@ public static <T, A extends Collection<T>> Collector<T, A, Collection<T>> toImmu
     /**
      * @param <T> The type
      * @return An immutable collection
+     * @deprecated use Stream#toList
      */
+    @Deprecated(forRemoval = true)
```
Comment (sdelamo) [inline core/src/main/java/io/micronaut/core/util/StreamUtils.java:196]: I think we should deprecate this in MN 3.9.x branch.

File: graal/src/main/java/io/micronaut/graal/reflect/GraalReflectionMetadataWriter.java [hunk 17]
```diff
@@ -40,7 +40,7 @@ final class GraalReflectionMetadataWriter extends AbstractAnnotationMetadataWrit
 
     public GraalReflectionMetadataWriter(ClassElement originatingElement,
                                          AnnotationMetadata annotationMetadata) {
-        super(resolveName(originatingElement), originatingElement, annotationMetadata, false);
+        super(resolveName(originatingElement), originatingElement, annotationMetadata, true);
```
Comment (sdelamo) [inline graal/src/main/java/io/micronaut/graal/reflect/GraalReflectionMetadataWriter.java:43]: its this change to use `true` as the constructor parameter intentional?

File: http-netty/src/main/java/io/micronaut/http/netty/channel/converters/KQueueChannelOptionFactory.java [hunk 18]
```diff
@@ -36,7 +36,7 @@
  * @author croudet
  */
 @Internal
-@Singleton
+@Prototype
```
Comment (sdelamo) [inline http-netty/src/main/java/io/micronaut/http/netty/channel/converters/KQueueChannelOptionFactory.java:39]: is the scope change intentional?

**Review by sdelamo — COMMENTED** [review]
Can you describe how are you measuring the performance improvements before and after this change?

File: context/src/main/java/io/micronaut/runtime/context/env/ConfigurationIntroductionAdvice.java (hunk 12 above)
Comment (dstepanov) [reply context/src/main/java/io/micronaut/runtime/context/env/ConfigurationIntroductionAdvice.java:109]: It was a visible point in the profiler, lamba is created and possibly invoked every time

File: core/src/main/java/io/micronaut/core/io/file/DefaultFileSystemResourceLoader.java (hunk 13 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/io/file/DefaultFileSystemResourceLoader.java:69]: I noticed it’s not needed to be nonempty as it doesn’t return a null or an optional

File: core/src/main/java/io/micronaut/core/type/Argument.java (hunk 14 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/type/Argument.java:439]: Yes, I have added new tests which were failing because that’s the only point providing decapitalized argument name

File: core/src/main/java/io/micronaut/core/type/DefaultArgument.java (hunk 15 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/type/DefaultArgument.java:235]: We need to have one way to provide the default argument name. Not really sure what should be the default

File: graal/src/main/java/io/micronaut/graal/reflect/GraalReflectionMetadataWriter.java (hunk 17 above)
Comment (dstepanov) [reply graal/src/main/java/io/micronaut/graal/reflect/GraalReflectionMetadataWriter.java:43]: Yes, the field is always true. The constructor is deprecated

File: gradle/libs.versions.toml (hunk 4 above)
Comment (dstepanov) [reply gradle/libs.versions.toml:179]: I converted all type registrars to be prototypes, now I see it also implements another interface, maybe it needs to be split

File: http-netty/src/main/java/io/micronaut/http/netty/channel/converters/KQueueChannelOptionFactory.java (hunk 18 above)
Comment (dstepanov) [reply http-netty/src/main/java/io/micronaut/http/netty/channel/converters/KQueueChannelOptionFactory.java:39]: I converted all type registrars to be prototypes, now I see it also implements another interface, maybe it needs to be split

File: buffer-netty/src/main/java/io/micronaut/buffer/netty/NettyByteBufferFactory.java [hunk 19]
```diff
@@ -42,20 +44,30 @@ public class NettyByteBufferFactory implements ByteBufferFactory<ByteBufAllocato
...
+        };
     }
 
     /**
-     * @param allocator The {@link ByteBufAllocator}
+     * @param allocatorSupplier The {@link ByteBufAllocator}
      */
-    public NettyByteBufferFactory(ByteBufAllocator allocator) {
-        this.allocator = allocator;
+    public NettyByteBufferFactory(ByteBufAllocator allocatorSupplier) {
```
Comment (yawkat) [inline buffer-netty/src/main/java/io/micronaut/buffer/netty/NettyByteBufferFactory.java]: param name is wrong

File: buffer-netty/src/main/java/io/micronaut/buffer/netty/NettyByteBufferFactory.java [hunk 20]
```diff
@@ -42,20 +44,30 @@ public class NettyByteBufferFactory implements ByteBufferFactory<ByteBufAllocato
...
 
-    private final ByteBufAllocator allocator;
+    private final Supplier<ByteBufAllocator> allocatorSupplier;
 
     /**
      * Default constructor.
      */
     public NettyByteBufferFactory() {
-        this.allocator = ByteBufAllocator.DEFAULT;
+        this.allocatorSupplier = new Supplier<ByteBufAllocator>() {
```
Comment (yawkat) [inline buffer-netty/src/main/java/io/micronaut/buffer/netty/NettyByteBufferFactory.java:53]: what's the point of this change? delaying initialization of ByteBufAllocator?

File: core/src/main/java/io/micronaut/core/convert/TypeConverter.java [hunk 21]
```diff
@@ -72,6 +72,12 @@ default Optional<T> convert(S object, Class<T> targetType) {
      * @return The converter instance
      */
     static <ST, TT> TypeConverter<ST, TT> of(Class<ST> sourceType, Class<TT> targetType, Function<ST, TT> converter) {
-        return (object, targetType1, context) -> Optional.ofNullable(converter.apply(object));
+        // Keep the anonymous class instead of Lambda to reduce the Lambda invocation overhead during the startup
```
Comment (yawkat) [inline core/src/main/java/io/micronaut/core/convert/TypeConverter.java:75]: i wonder if we could use retrolambda to do these changes automatically. but maybe it's overkill.

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java [hunk 22]
```diff
@@ -4103,48 +4090,102 @@ public <T> BeanRegistration<T> getInFlightBean(BeanIdentifier beanIdentifier) {
...
 
-        @Override
-        protected void onNewSegment(Segment<?, ?> segment) {
-            Argument<?> argument = segment.getArgument();
-            if (argument.isContainerType()) {
-                argument = argument.getFirstTypeVariable().orElse(null);
-                if (argument == null) {
-                    return;
+        public boolean isDisabled() {
+            return reference == null || referenceEnabled != null && !referenceEnabled || definitionEnabled != null && !definitionEnabled;
```
Comment (yawkat) [inline inject/src/main/java/io/micronaut/context/DefaultBeanContext.java]: add parentheses, this is unreadable

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java [hunk 23]
```diff
@@ -4103,48 +4090,102 @@ public <T> BeanRegistration<T> getInFlightBean(BeanIdentifier beanIdentifier) {
...
-                            .add(getHierarchy());
+            return reference;
+        }
+
+        public <T> BeanDefinition<T> getDefinition(BeanContext beanContext) {
+            if (definitionEnabled != null && !definitionEnabled) {
+                throw new IllegalStateException("The definition is not enabled");
+            }
+            try {
+                if (definition == null) {
```
Comment (yawkat) [inline inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:4177]: this is racey with isDefinitionEnabled

File: inject/src/main/java/io/micronaut/inject/BeanDefinition.java (hunk 10 above)
Comment (yawkat) [reply inject/src/main/java/io/micronaut/inject/BeanDefinition.java]: `?` is correct though? it would be better to change downstream users of `getExecutableMethods` to use proper covariance

tbh maybe we could just drop the return type parameter from ExecutableMethod, i dont think we use it productively anyway

**Review by yawkat — APPROVED** [review]

File: buffer-netty/src/main/java/io/micronaut/buffer/netty/NettyByteBufferFactory.java (hunk 20 above)
Comment (dstepanov) [reply buffer-netty/src/main/java/io/micronaut/buffer/netty/NettyByteBufferFactory.java:53]: Yes, postpone it to a bit later

**Comment by dstepanov** [comment]
I removed the changes to `getExecutableMethods`. I'll add them in another PR. Removing the return generic type might be a solution.

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java (hunk 23 above)
Comment (dstepanov) [reply inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:4177]: We don't really care, it can be loaded multiple times like it's now
